crypto/rsa/rsa_ossl.c

```diff
@@ -664,10 +668,37 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     }
 
     if (rsa->flags & RSA_FLAG_CACHE_PUBLIC)
-        if (!BN_MONT_CTX_set_locked
-            (&rsa->_method_mod_n, rsa->lock, rsa->n, ctx))
+        if (!BN_MONT_CTX_set_locked(&rsa->_method_mod_n, rsa->lock,
+                                    rsa->n, ctx))
+            goto err;
+
+    if (smooth) {
+        if (/* m1 = I moq q */
+            !bn_from_mont_fixed_top(m1, I, rsa->_method_mod_q, ctx)
+            || !bn_to_mont_fixed_top(m1, m1, rsa->_method_mod_q, ctx)
+            /* m1 = m1^dmq1 mod q */
+            || !BN_mod_exp_mont_consttime(m1, m1, rsa->dmq1, rsa->q, ctx,
+                                          rsa->_method_mod_q)
+            /* r1 = I mod p */
+            || !bn_from_mont_fixed_top(r1, I, rsa->_method_mod_p, ctx)
+            || !bn_to_mont_fixed_top(r1, r1, rsa->_method_mod_p, ctx)
+            /* r1 = r1^dmp1 mod p */
+            || !BN_mod_exp_mont_consttime(r1, r1, rsa->dmp1, rsa->p, ctx,
+                                          rsa->_method_mod_p)
+            /* r1 = (r1 - m1) mod p */
```

Comment:
This is probably worth elaborating on in the comment since it's not obvious. While `r1` is fully reduced mod `p`, `m1` is only fully reduced mod `q`, which may be larger than `p`. This doesn't fit in the "obvious" modular subtraction precondition, but `bn_mod_sub_fixed_top` does explicitly allows for this.

(We instead canonicalized things so that p > q, which is differently complicated; we end up storing extra copies of several fields to deal with this and other input width issues. But we also removed `RSA_FLAG_CACHE_PRIVATE` and always cache stuff, so that option was easier for us. But it did allow a minor optimization: cache iqmp in Montgomery form to save a Montgomery reduction. I should note that the more general `bn_mod_sub_fixed_top` is then less efficient for the case when both are guaranteed to be reduced. Notably, the EC code needs a lot of modular subtractions, when it comes time to fix its timing issues.)

Reply:
> This doesn't fit in the "obvious" modular subtraction precondition, but bn_mod_sub_fixed_top does explicitly allows for this.

Just in case, I did verify q>p to work. I'll add commentary [tomorrow]...